### PR TITLE
Improve Recording Options in menu

### LINKED ISSUE #86: Make multiplayer name configurable in settings

### LINKED ISSUE #87: Make multiplayer name configurable in 'Save Recording' menu

### LINKED ISSUE #88: Save / Load / Delete Recording Menus
Completely redo 'Save Recording' menu with the following:
---

- [x] An input field for a file name at the top and one button per existing recording below. [FORMER IS SELF-EXPLANATORY, FOR THE LATTER YOU NEED TO ITERATE OVER EACH ITEM IN THE LIST (for recording in recordings_list { }) and then make buttons for each recording within that iteration block]

- [x] If the input field is submitted, then the recording of the input field name is taken. [DO EXISTENCE CHECK OF THE INPUT IN THE RECORDING DIR]

- [x] If a button of an existing recording is pressed, then that recording is overwritten. [ANOTHER EXISTENCE CHECK, IF IT EXISTS, OVERWRITES IT. IF IT DOESN'T EXIST, DO THE NORMAL STUFF]

- [x] For Load Recording and Delete Recording I'd only show one button per recording without an input field (or with an input field for olivia as the list of existing recordings may get too long).

### PR DESCRIPTION
**Pull request by LukeSaward1**
Previous discussion: #85

Fix #86, fix #87, fix #88

### CODE REVIEW

File: tool/settings.re [hunk 1]
```diff
@@ -68,7 +68,12 @@ impl Settings {
         self.store();
     }
     fn set_multiplayer_name(mut self, name: string) {
+        print(f"1:   name: {name}\nself.multiplayer_name: {self.multiplayer_name}\nSETTINGS.multiplayer_name: {SETTINGS.multiplayer_name}");
+        SETTINGS.multiplayer_name = name;
+        print(f"2:   name: {name}\nself.multiplayer_name: {self.multiplayer_name}\nSETTINGS.multiplayer_name: {SETTINGS.multiplayer_name}");
         self.multiplayer_name = name;
+        print(f"3:   name: {name}\nself.multiplayer_name: {self.multiplayer_name}\nSETTINGS.multiplayer_name: {SETTINGS.multiplayer_name}");
         self.store();
+        print(f"4:   name: {name}\nself.multiplayer_name: {self.multiplayer_name}\nSETTINGS.multiplayer_name: {SETTINGS.multiplayer_name}");
```
Comment (oberien) [inline tool/settings.re:77]: debugging still left in

File: tool/main.re [hunk 2]
```diff
@@ -246,22 +250,188 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
     }),
 ));
 
+static mut RECORDING_NAME_LABEL = Text { text: "Recording name" };
+
+fn recording_operations(operation: string) {
```
Comment (oberien) [inline tool/main.re:255]: This function is unused and the internal code is now quadruplicated.

File: tool/main.re [hunk 3]
```diff
@@ -246,22 +250,188 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+                if recordings_list.contains(input) && operation == "load" {
+                    RECORDING_NAME_LABEL.text = f"Recording name";
+                    set_current_component(TAS_COMPONENT);
+                    leave_ui();
+                    leave_ui();
+                    leave_ui();
+                } else {
+                    RECORDING_NAME_LABEL.text = f"Recording name (Error: no such recording)";
+                }
+                if recordings_list.contains(input) && operation == "delete" {
```
Comment (oberien) [inline tool/main.re:281]: Instead of using string-based operations, create an enum defining all possible operations.

File: tool/main.re [hunk 4]
```diff
@@ -246,22 +250,188 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+                    print("Save");
+                    tas_save_recording(input);
+                }
+            },
+            onchange: fn(input: string) {}
+        }),
+    );
+}
+
+fn handle_operation(op: string, input: string) {
```
Comment (oberien) [inline tool/main.re:294]: This function only overcomplicates things. It refactors out things that are not actually duplicated but unique to each operation.

**Review by oberien — CHANGES_REQUESTED** [review]

File: tool/main.re [hunk 5]
```diff
@@ -246,22 +250,143 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+    let mut recordings = List::of(
+        UiElement::Button(UiButton {
+            label: Text { text: "Back" },
+            onclick: fn(label: Text) { leave_ui() },
+        }),
+        UiElement::Input(Input {
+            label: RECORDING_NAME_LABEL,
+            input: "",
+            onclick: fn(input: string) {
+                let recordings_list = Tas::list_recordings();
```
Comment (oberien) [inline tool/main.re:274]: unused variable

File: tool/main.re [hunk 6]
```diff
@@ -246,22 +250,143 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+                        leave_ui();
+                        leave_ui();
+                        leave_ui();
+                    },
+                    ReplayButtonOp::Delete => {
+                        Tas::remove_recording(input);
+                        leave_ui();
+                    },
+                    _ => panic(f"unreachable: got ReplayButtonOp {op}"),
+                }
```
Comment (oberien) [inline tool/main.re:295]: Refactor this out into a closure.

File: tool/main.re [hunk 7]
```diff
@@ -246,22 +250,143 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+                        leave_ui();
+                        leave_ui();
+
+                    },
+                    ReplayButtonOp::Delete => {
+                        Tas::remove_recording(label.text);
+                        leave_ui();
+                    },
+                    _ => panic(f"unreachable: got op {op}"),
+                }
```
Comment (oberien) [inline tool/main.re:322]: reuse the above closure here

File: tool/main.re [hunk 8]
```diff
@@ -246,22 +250,143 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+                    label: Text { text: recording },
+                    onclick: fn(label: Text) {
+                        Tas::remove_recording(label.text); // The list can only update when you re-enter the menu
+                        RECORDING_NAME_LABEL.text = f"Recording name (Deleted {label.text})";
+                    },
+                }));
+            }
+            enter_ui(Ui::new("Recording Options:", recordings));
+        }
     }),
```
Comment (oberien) [inline tool/main.re:390]: Why is this still here?

**Review by oberien — CHANGES_REQUESTED** [review]

File: tool/main.re [hunk 9]
```diff
@@ -246,22 +250,143 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
 
+static mut RECORDING_NAME_LABEL = Text { text: "Recording name" };
+
+enum ReplayButtonOp {
+    Save,
+    Load,
+    Delete,
+}
+
+fn handle_operation(op: ReplayButtonOp) {
```
Comment (oberien) [inline tool/main.re:261]:
```suggestion
fn handle_replay_op(op: ReplayButtonOp) {
```

File: tool/main.re (hunk 8 above)
Comment (LukeSaward1) [reply tool/main.re:390]: LMFAO I didn't realize this was still in here

File: tool/main.re [hunk 10]
```diff
@@ -246,22 +250,100 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+static mut RECORDING_NAME_LABEL = Text { text: "Recording name" };
+
+enum ReplayButtonOp {
+    Save,
+    Load,
+    Delete,
+}
+
+fn handle_replay_op(op: ReplayButtonOp) {
+    set_current_component(TAS_COMPONENT);
```
Comment (oberien) [inline tool/main.re:262]: This is only needed when successfully loading a recording.

File: tool/main.re [hunk 11]
```diff
@@ -246,22 +250,100 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+            },
+            ReplayButtonOp::Load => {
+                if recordings_list.contains(input) {
+                    tas_load_recording(input);
+                    leave_ui();
+                    leave_ui();
+                    leave_ui();
+                } else {
+                    RECORDING_NAME_LABEL.text = f"Recording name (Error: no such file)";
+                }
```
Comment (oberien) [inline tool/main.re:278]: Use a guard-if instead.

File: tool/main.re [hunk 12]
```diff
@@ -246,22 +250,100 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
+            },
+            ReplayButtonOp::Delete => {
+                if recordings_list.contains(input) {
+                    Tas::remove_recording(input);
+                    leave_ui();
+                } else {
+                    RECORDING_NAME_LABEL.text = f"Recording name (Error: no such file)";
+                }
```
Comment (oberien) [inline tool/main.re:286]: ~

File: tool/main.re [hunk 13]
```diff
@@ -246,22 +250,100 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
...
 
+static mut RECORDING_NAME_LABEL = Text { text: "Recording name" };
+
+enum ReplayButtonOp {
+    Save,
+    Load,
+    Delete,
+}
+
+fn handle_replay_op(op: ReplayButtonOp) {
```
Comment (oberien) [inline tool/main.re:261]:
```suggestion
fn enter_replay_menu(op: ReplayButtonOp) {
```

File: tool/main.re [hunk 14]
```diff
@@ -246,22 +250,100 @@ static MULTIPLAYER_MENU = Ui::new("Multiplayer:", List::of(
     }),
 ));
 
+static mut RECORDING_NAME_LABEL = Text { text: "Recording name" };
+
+enum ReplayButtonOp {
```
Comment (oberien) [inline tool/main.re:255]:
```suggestion
enum ReplayMenu {
```

**Review by oberien — CHANGES_REQUESTED** [review]

**Review by oberien — APPROVED** [review]
Ta for the PR.

**Pull request merged**